### ISSUE TITLE
gpu memory consumption increases when there is no softmax loss.

### ISSUE THREAD

**Issue by leon-liangwu** [NVIDIA/caffe#474]
I notice that version 0.16.5 consume more gpu memory without softmax loss.
Does anyone find this too?

**Comment by drnikolaev** [comment]
This would be helpful to see some logs confirming the statement.

**Comment by leon-liangwu** [comment]
@drnikolaev 
This is the gpu memory consumption when I train the classification target.
![mb-reg-0 75-loss](https://user-images.githubusercontent.com/7507360/35679690-fbeb3a14-0792-11e8-9684-947037fa1481.png)
The prototxt is here. 
[MobileNet_deploy_0.75.prototxt.txt](https://github.com/NVIDIA/caffe/files/1685600/MobileNet_deploy_0.75.prototxt.txt)


If I delete the loss leaving the conv to be the last layer. The gpu memory increases dramatically.
![mb-reg-0 75-no_loss](https://user-images.githubusercontent.com/7507360/35679937-c563d23e-0793-11e8-9549-63cfef3abdf8.png)
The prototxt is here.
[MobileNet_deploy_0.75-no_loss.prototxt.txt](https://github.com/NVIDIA/caffe/files/1685601/MobileNet_deploy_0.75-no_loss.prototxt.txt)

Could you please me to fix this?

**Comment by drnikolaev** [comment]
@DreamHorse0301 this is correct behavior. Have you noticed that instead of having single output like this:
```
I0201 13:58:13.857707 17831 solver.cpp:371]     Train net output #0: loss = 6.19141 (* 1 = 6.19141 loss)
```
you get 64064 ones like this:
```
...
I0201 14:01:56.670811 18036 solver.cpp:371]     Train net output #63997: fc7 = -0.168701
I0201 14:01:56.670816 18036 solver.cpp:371]     Train net output #63998: fc7 = 0.0449829
I0201 14:01:56.670823 18036 solver.cpp:371]     Train net output #63999: fc7 = 0.330322
I0201 14:01:56.671797 18036 solver.cpp:371]     Train net output #64000: label = 662
I0201 14:01:56.671815 18036 solver.cpp:371]     Train net output #64001: label = 697
I0201 14:01:56.671823 18036 solver.cpp:371]     Train net output #64002: label = 290
...
```
You removed aggregating layer and now NVCaffe has to maintain them all. It takes GPU memory. See also this code:
```C++
  // In the end, all remaining blobs are considered output blobs.
  for (set<string>::iterator it = available_blobs.begin();
      it != available_blobs.end(); ++it) {
    LOG_IF(INFO, Caffe::root_solver())
        << "This network produces output " << *it;
    net_output_blobs_.push_back(blobs_[blob_name_to_idx[*it]].get());
    net_output_blob_indices_.push_back(blob_name_to_idx[*it]);
  }
```
